ProcessQueue/__init__.py

```diff
@@ -180,37 +190,57 @@ def perform_chunking(entries_to_post):
     return chunked_entries_to_post
 
 
-# @memory_profiler.profile(stream=profiler_logstream)
-def paginate(entries_to_post, max_chars=None, other=""):
+def paginate(entries, max_chars=None):
     """
-    This expects a list of dicts, and returns a list of lists of dicts,
-    where the maximum length of each list of dicts, under JSONification,
+    This expects a list of strings, and returns a list of lists of strings,
+    where the maximum length of each list of strings, under JSONification,
     is less than max_chars
     """
-    if not max_chars:
-        max_chars = (
-            int(os.environ.get("PAGE_MAX_CHARS"))
-            if os.environ.get("PAGE_MAX_CHARS")
-            else 10000
-        )
-    max_chars = max_chars - len(other)
+    max_chars = handle_max_chars(max_chars)
 
-    paginated_entries_to_post = []
+    paginated_entries = []
     this_page = []
-    for entry in entries_to_post:
+    for entry in entries:
         # If the current page won't be overfull, add the entry to the current page
         if len(json.dumps(this_page)) + len(json.dumps(entry)) < max_chars:
             this_page.append(entry)
         else:
             # Otherwise, this page should be added to the list of pages.
-            paginated_entries_to_post.append(this_page)
+            paginated_entries.append(this_page)
             # Now add the entry that would have over-filled the page.
             this_page = [entry]
 
     # After all entries are added, check for a half-filled page, and if present add it to the list of pages
     if this_page:
-        paginated_entries_to_post.append(this_page)
-    return paginated_entries_to_post
+        paginated_entries.append(this_page)
+
+    return paginated_entries
+
+
+def paginate_two_lists(entries, other, max_chars=None):
+    """
+    This expects two lists of strings, and returns a list of 2-tuples of lists of
+    strings, where the maximum length of each tuple of list of strings, under
+    JSONification, is less than max_chars
+
+    An optimum strategy to minimise the number of returned tuples would be complex. Here
+    we use a simple heuristic: if the two lists fit together under the limit,
+    then return them as one tuple. Otherwise, split each list to individually fit
+    into max_chars/2, and recombine the product of the two into a list of tuples.
+    """
+    max_chars = handle_max_chars(max_chars)
+
+    # If the two lists are short enough, then return a single tuple
+    if len(json.dumps(entries)) + len(json.dumps(other)) < max_chars:
+        return [(entries, other)]
+
+    paginated_entries = paginate(entries, max_chars / 2)
+    paginated_other = paginate(other, max_chars / 2)
+    return [
+        (page_entries, page_other)
+        for page_entries in paginated_entries
+        for page_other in paginated_other
+    ]
```

Comment:
I'm not sure about this statement. Isn't this like doing:
```Python
for page_entries in paginated_entries:
    for page_other in paginated_other:
        # do stuff
```
resulting in a `len(paginated_entries)` x `len(paginated_others)` array?